<a href="https://colab.research.google.com/github/bharathulaprasad/AprioriAlgorithm/blob/main/YetYetAnotherApriori.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

We have the following transaction data:
t1: Beef, Chicken, Milk
t2: Beef, Cheese
t3: Cheese, Boots
t4: Beef, Chicken, Cheese
t5: Beef, Chicken, Clothes, Cheese, Milk
t6: Chicken, Clothes, Milk
t7: Chicken, Milk, Clothes

In [1]:
#t1: Beef, Chicken, Milk
#t2: Beef, Cheese
#t3: Cheese, Boots
#t4: Beef, Chicken, Cheese
#t5: Beef, Chicken, Clothes, Cheese, Milk
#t6: Chicken, Clothes, Milk
#t7: Chicken, Milk, Clothes

dataset=[['Beef', 'Chicken', 'Milk'],
         ['Beef', 'Cheese'],
         ['Cheese', 'Boots'],
         ['Beef', 'Chicken', 'Cheese'],
         ['Beef', 'Chicken', 'Clothes', 'Cheese', 'Milk'],
         ['Chicken', 'Clothes', 'Milk'],
         ['Chicken', 'Milk', 'Clothes']]

# Q1) What would be the confidence of the following rule:

## Beef, Cheese --> Cheese

In [2]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder

In [3]:
dataset

[['Beef', 'Chicken', 'Milk'],
 ['Beef', 'Cheese'],
 ['Cheese', 'Boots'],
 ['Beef', 'Chicken', 'Cheese'],
 ['Beef', 'Chicken', 'Clothes', 'Cheese', 'Milk'],
 ['Chicken', 'Clothes', 'Milk'],
 ['Chicken', 'Milk', 'Clothes']]

In [4]:
te= TransactionEncoder()
te_ary=te.fit(dataset).transform(dataset)

In [5]:
df=pd.DataFrame(te_ary,columns=te.columns_)
df

Beef  Boots  Cheese  Chicken  Clothes   Milk
0   True  False   False     True    False   True
1   True  False    True    False    False  False
2  False   True    True    False    False  False
3   True  False    True     True    False  False
4   True  False    True     True     True   True
5  False  False   False     True     True   True
6  False  False   False     True     True   True

#Find frequent itemsets

In [6]:
from mlxtend import frequent_patterns
from mlxtend.frequent_patterns import apriori
frequent_itemsets=apriori(df,min_support=0.1,use_colnames=True)
frequent_itemsets

support                                itemsets
0   0.571429                                  (Beef)
1   0.142857                                 (Boots)
2   0.571429                                (Cheese)
3   0.714286                               (Chicken)
4   0.428571                               (Clothes)
5   0.571429                                  (Milk)
6   0.428571                          (Cheese, Beef)
7   0.428571                         (Chicken, Beef)
8   0.142857                         (Beef, Clothes)
9   0.285714                            (Milk, Beef)
10  0.142857                         (Cheese, Boots)
11  0.285714                       (Chicken, Cheese)
12  0.142857                       (Cheese, Clothes)
13  0.142857                          (Cheese, Milk)
14  0.428571                      (Chicken, Clothes)
15  0.571429                         (Chicken, Milk)
16  0.428571                         (Milk, Clothes)
17  0.285714                 (Chicken, Cheese, Beef)
18  0.142857                 (Cheese, Beef, Clothes)
19  0.142857                    (Cheese, Milk, Beef)
20  0.142857                (Chicken, Beef, Clothes)
21  0.285714                   (Chicken, Milk, Beef)
22  0.142857                   (Milk, Beef, Clothes)
23  0.142857              (Chicken, Cheese, Clothes)
24  0.142857                 (Chicken, Cheese, Milk)
25  0.142857                 (Cheese, Milk, Clothes)
26  0.428571                (Chicken, Milk, Clothes)
27  0.142857        (Chicken, Cheese, Beef, Clothes)
28  0.142857           (Chicken, Cheese, Milk, Beef)
29  0.142857           (Cheese, Milk, Beef, Clothes)
30  0.142857          (Chicken, Milk, Beef, Clothes)
31  0.142857        (Chicken, Cheese, Milk, Clothes)
32  0.142857  (Chicken, Clothes, Beef, Milk, Cheese)

In [7]:
from mlxtend.frequent_patterns import association_rules
res = association_rules(frequent_itemsets,metric='confidence')
res

antecedents  ... conviction
0                            (Boots)  ...        inf
1                          (Clothes)  ...        inf
2                             (Milk)  ...        inf
3                          (Clothes)  ...        inf
4                  (Chicken, Cheese)  ...        inf
5                  (Cheese, Clothes)  ...        inf
6                    (Beef, Clothes)  ...        inf
7                     (Cheese, Milk)  ...        inf
8                    (Beef, Clothes)  ...        inf
9                       (Milk, Beef)  ...        inf
10                   (Beef, Clothes)  ...        inf
11                 (Cheese, Clothes)  ...        inf
12                    (Cheese, Milk)  ...        inf
13                    (Cheese, Milk)  ...        inf
14                 (Cheese, Clothes)  ...        inf
15                (Chicken, Clothes)  ...        inf
16                   (Milk, Clothes)  ...        inf
17                         (Clothes)  ...        inf
18        (Chicken, Clothes, Cheese)  ...        inf
19          (Chicken, Beef, Clothes)  ...        inf
20           (Cheese, Beef, Clothes)  ...        inf
21                 (Cheese, Clothes)  ...        inf
22                   (Beef, Clothes)  ...        inf
23           (Chicken, Milk, Cheese)  ...        inf
24              (Cheese, Milk, Beef)  ...        inf
25                    (Cheese, Milk)  ...        inf
26              (Cheese, Milk, Beef)  ...        inf
27           (Cheese, Milk, Clothes)  ...        inf
28           (Cheese, Beef, Clothes)  ...        inf
29             (Milk, Beef, Clothes)  ...        inf
30                    (Cheese, Milk)  ...        inf
31                 (Cheese, Clothes)  ...        inf
32                   (Beef, Clothes)  ...        inf
33          (Chicken, Beef, Clothes)  ...        inf
34             (Milk, Beef, Clothes)  ...        inf
35                   (Beef, Clothes)  ...        inf
36           (Chicken, Milk, Cheese)  ...        inf
37        (Chicken, Clothes, Cheese)  ...        inf
38           (Cheese, Milk, Clothes)  ...        inf
39                    (Cheese, Milk)  ...        inf
40                 (Cheese, Clothes)  ...        inf
41    (Chicken, Milk, Beef, Clothes)  ...        inf
42  (Chicken, Beef, Clothes, Cheese)  ...        inf
43  (Chicken, Milk, Clothes, Cheese)  ...        inf
44     (Chicken, Milk, Beef, Cheese)  ...        inf
45     (Cheese, Milk, Beef, Clothes)  ...        inf
46          (Chicken, Beef, Clothes)  ...        inf
47        (Chicken, Clothes, Cheese)  ...        inf
48           (Chicken, Milk, Cheese)  ...        inf
49             (Milk, Beef, Clothes)  ...        inf
50           (Cheese, Beef, Clothes)  ...        inf
51           (Cheese, Milk, Clothes)  ...        inf
52              (Cheese, Milk, Beef)  ...        inf
53                   (Beef, Clothes)  ...        inf
54                 (Cheese, Clothes)  ...        inf
55                    (Cheese, Milk)  ...        inf

[56 rows x 9 columns]

So answer is 0.

## { Beef, Cheese, Chicken} can be called as frequent itemset if the support threshold is set to 0.2 (True/False)

In [8]:
from mlxtend import frequent_patterns
from mlxtend.frequent_patterns import apriori
frequent_itemsets=apriori(df,min_support=0.2,use_colnames=True)
frequent_itemsets

support                  itemsets
0   0.571429                    (Beef)
1   0.571429                  (Cheese)
2   0.714286                 (Chicken)
3   0.428571                 (Clothes)
4   0.571429                    (Milk)
5   0.428571            (Cheese, Beef)
6   0.428571           (Chicken, Beef)
7   0.285714              (Milk, Beef)
8   0.285714         (Chicken, Cheese)
9   0.428571        (Chicken, Clothes)
10  0.571429           (Chicken, Milk)
11  0.428571           (Milk, Clothes)
12  0.285714   (Chicken, Cheese, Beef)
13  0.285714     (Chicken, Milk, Beef)
14  0.428571  (Chicken, Milk, Clothes)

TRUE

What would be the support of the following rule:
Beef, Cheese --> Chicken

Answer: 0.285714

In [10]:
from mlxtend.frequent_patterns import association_rules
res = association_rules(frequent_itemsets,metric='confidence',min_threshold=0.1)
res

antecedents         consequents  ...  leverage  conviction
0             (Cheese)              (Beef)  ...  0.102041    1.714286
1               (Beef)            (Cheese)  ...  0.102041    1.714286
2            (Chicken)              (Beef)  ...  0.020408    1.071429
3               (Beef)           (Chicken)  ...  0.020408    1.142857
4               (Milk)              (Beef)  ... -0.040816    0.857143
5               (Beef)              (Milk)  ... -0.040816    0.857143
6            (Chicken)            (Cheese)  ... -0.122449    0.714286
7             (Cheese)           (Chicken)  ... -0.122449    0.571429
8            (Chicken)           (Clothes)  ...  0.122449    1.428571
9            (Clothes)           (Chicken)  ...  0.122449         inf
10           (Chicken)              (Milk)  ...  0.163265    2.142857
11              (Milk)           (Chicken)  ...  0.163265         inf
12              (Milk)           (Clothes)  ...  0.183673    2.285714
13           (Clothes)              (Milk)  ...  0.183673         inf
14   (Chicken, Cheese)              (Beef)  ...  0.122449         inf
15     (Chicken, Beef)            (Cheese)  ...  0.040816    1.285714
16      (Cheese, Beef)           (Chicken)  ... -0.020408    0.857143
17           (Chicken)      (Cheese, Beef)  ... -0.020408    0.952381
18            (Cheese)     (Chicken, Beef)  ...  0.040816    1.142857
19              (Beef)   (Chicken, Cheese)  ...  0.122449    1.428571
20     (Chicken, Milk)              (Beef)  ... -0.040816    0.857143
21     (Chicken, Beef)              (Milk)  ...  0.040816    1.285714
22        (Milk, Beef)           (Chicken)  ...  0.081633         inf
23           (Chicken)        (Milk, Beef)  ...  0.081633    1.190476
24              (Milk)     (Chicken, Beef)  ...  0.040816    1.142857
25              (Beef)     (Chicken, Milk)  ... -0.040816    0.857143
26     (Chicken, Milk)           (Clothes)  ...  0.183673    2.285714
27  (Chicken, Clothes)              (Milk)  ...  0.183673         inf
28     (Milk, Clothes)           (Chicken)  ...  0.122449         inf
29           (Chicken)     (Milk, Clothes)  ...  0.122449    1.428571
30              (Milk)  (Chicken, Clothes)  ...  0.183673    2.285714
31           (Clothes)     (Chicken, Milk)  ...  0.183673         inf

[32 rows x 9 columns]

In [12]:
# We can derive the following rule if we set the minimum support and minimum confidence - 50% (True/False)
# Beef, Cheese --> Chicken
# Answer: False
from mlxtend import frequent_patterns
from mlxtend.frequent_patterns import apriori
frequent_itemsets=apriori(df,min_support=0.5,use_colnames=True)
from mlxtend.frequent_patterns import association_rules
res = association_rules(frequent_itemsets,metric='confidence',min_threshold=0.5)
res

antecedents consequents  antecedent support  ...  lift  leverage  conviction
0   (Chicken)      (Milk)            0.714286  ...   1.4  0.163265    2.142857
1      (Milk)   (Chicken)            0.571429  ...   1.4  0.163265         inf

[2 rows x 9 columns]

In [ ]:
#What would be the confidence of the following rule:
#Beef --> Milk
#Answer: 0.5

In [14]:
#What would be the confidence of the following rule:
#Cheese --> Boots
from mlxtend import frequent_patterns
from mlxtend.frequent_patterns import apriori
frequent_itemsets=apriori(df,min_support=0.1,use_colnames=True)

from mlxtend.frequent_patterns import association_rules
res = association_rules(frequent_itemsets,metric='confidence',min_threshold=0.1)
res
# Answer: 0.25

antecedents                       consequents  ...  leverage  conviction
0      (Cheese)                            (Beef)  ...  0.102041    1.714286
1        (Beef)                          (Cheese)  ...  0.102041    1.714286
2     (Chicken)                            (Beef)  ...  0.020408    1.071429
3        (Beef)                         (Chicken)  ...  0.020408    1.142857
4        (Beef)                         (Clothes)  ... -0.102041    0.761905
..          ...                               ...  ...       ...         ...
177   (Chicken)     (Cheese, Milk, Beef, Clothes)  ...  0.040816    1.071429
178   (Clothes)     (Chicken, Milk, Beef, Cheese)  ...  0.081633    1.285714
179      (Beef)  (Chicken, Milk, Clothes, Cheese)  ...  0.061224    1.142857
180      (Milk)  (Chicken, Beef, Clothes, Cheese)  ...  0.061224    1.142857
181    (Cheese)    (Chicken, Milk, Beef, Clothes)  ...  0.061224    1.142857

[182 rows x 9 columns]

{Beef, Cheese, Chicken} can be called as frequent itemset if the support threshold is set to 0.3 (True/False)
Answer False. It can be True if support threshold is 0.28

In [20]:
from mlxtend import frequent_patterns
from mlxtend.frequent_patterns import apriori
frequent_itemsets=apriori(df,min_support=0.28,use_colnames=True)
frequent_itemsets

support                  itemsets
0   0.571429                    (Beef)
1   0.571429                  (Cheese)
2   0.714286                 (Chicken)
3   0.428571                 (Clothes)
4   0.571429                    (Milk)
5   0.428571            (Cheese, Beef)
6   0.428571           (Chicken, Beef)
7   0.285714              (Milk, Beef)
8   0.285714         (Chicken, Cheese)
9   0.428571        (Chicken, Clothes)
10  0.571429           (Chicken, Milk)
11  0.428571           (Milk, Clothes)
12  0.285714   (Chicken, Cheese, Beef)
13  0.285714     (Chicken, Milk, Beef)
14  0.428571  (Chicken, Milk, Clothes)